### PR TITLE
Add conversion functions for Lon migration 

### PR DESCRIPTION
**Pull request by zeqing-guo**
1. Add `enableConversion`, `disableConversion`, and `convertXLonToVeLon` functions in lonStaking.sol, which helps xLon holders migrate to veLon directly;
2. Add `enableConversion`, `disableConversion`, and `convertVeLontoXXXLon` functions in veLon.sol to help veLon holders migrate to xxxLon directly in the future;
3. Write some test cases

### CODE REVIEW

File: contracts/LONStaking.sol [hunk 1]
```diff
@@ -294,12 +315,24 @@ contract LONStaking is ERC20ForUpgradeable, OwnableForUpgradeable, ReentrancyGua
...
-            _redeem(share.sub(penalty), penalty);
+            _redeemAndSendTo(share.sub(penalty), penalty, msg.sender);
         }
     }
+
+    function convertXLonToVeLon(uint256 _lockDuration) public returns (uint256 tokenId) {
+        require(conversion, "conversion is not enabled");
+
+        uint256 userShare = balanceOf(msg.sender);
+        uint256 userLonBalance = userShare.mul(lonToken.balanceOf(address(this))).div(totalSupply());
```
Comment (charlesjhongc) [inline contracts/LONStaking.sol:331]: How about return this amount from `_redeemAndSendTo()` to save some gas (reading storage)?

File: contracts/LONStaking.sol [hunk 2]
```diff
@@ -294,12 +315,24 @@ contract LONStaking is ERC20ForUpgradeable, OwnableForUpgradeable, ReentrancyGua
...
+        _redeemAndSendTo(userShare, 0, address(this));
+        IveLON veLonIntance = IveLON(veLon);
```
Comment (charlesjhongc) [inline contracts/LONStaking.sol:333]: `instance`? perhaps just `veLon`?

**Review by charlesjhongc — COMMENTED** [review]
Maybe have an interface includes `enable()` `disable()` and `migrate()` then make `LONStaking` and `veLON` inherit it?

File: contracts/LONStaking.sol (hunk 1 above)
Comment (zeqing-guo) [reply contracts/LONStaking.sol]: Good idea. I have added `returns (uint256 userTotalAmount)` for `_redeemAndSendTo()`.

File: contracts/LONStaking.sol (hunk 2 above)
Comment (zeqing-guo) [reply contracts/LONStaking.sol]: Done.

File: contracts/test/veLON/Convert.t.sol [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        assertEq(veLon.earlyWithdrawPenaltyRate(), penaltyRateBefore);
+    }
+
+    function testConvertVeLontoXXXLon() public {
+        uint256 tokenId = _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, DEFAULT_LOCK_TIME);
+        require(tokenId != 0, "No lock created yet");
+        _convertVeLontoXXXLonAndValidate(user, DEFAULT_STAKE_AMOUNT);
+    }
+
+    function _convertVeLontoXXXLonAndValidate(address staker, uint256 stakeAmount) public {
```
Comment (charlesjhongc) [inline contracts/test/veLON/Convert.t.sol:40]: internal?

File: contracts/test/veLON/Convert.t.sol [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        veLon.disableConversion();
+
+        assertEq(veLon.dstToken(), address(0x0));
+        assertEq(veLon.conversion(), false);
+        assertEq(veLon.earlyWithdrawPenaltyRate(), penaltyRateBefore);
+    }
+
+    function testConvertVeLontoXXXLon() public {
+        uint256 tokenId = _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, DEFAULT_LOCK_TIME);
+        require(tokenId != 0, "No lock created yet");
```
Comment (charlesjhongc) [inline contracts/test/veLON/Convert.t.sol:36]: use `assertGt` ?

File: contracts/test/LONStaking.t.sol [hunk 5]
```diff
@@ -1156,6 +1159,95 @@ contract LONStakingTest is Test {
...
+        vm.prank(user);
+        uint256 tokenId = lonStaking.convert(abi.encode(convertDuration));
+        uint256 expectedPower = _calcPower(convertDuration, convertAmount, veLon.maxLockDuration());
+
+        veLonLon.assertChange(int256(convertAmount));
+        lonStakingLon.assertChange(-int256(convertAmount));
+        userXLon.assertChange(-int256(userXLonAmount));
+
+        assertGe(tokenId, 0);
+        uint256 actualPower = veLon.vBalanceOf(tokenId);
```
Comment (charlesjhongc) [inline contracts/test/LONStaking.t.sol:1206]: `assertEq(veLon.vBalanceOf(tokenId), expectedPower);`
`assertEq(veLon.totalSupply(), expectedPower);`

File: contracts/test/LONStaking.t.sol [hunk 6]
```diff
@@ -1156,6 +1159,95 @@ contract LONStakingTest is Test {
...
+
+    function _convertXLonToVeLonAndValidate(uint256 convertAmount, uint256 convertDuration) internal {
+        _stake(user, convertAmount);
+        veLON veLon = new veLON(address(lon));
+
+        BalanceSnapshot.Snapshot memory veLonLon = BalanceSnapshot.take(address(veLon), address(lon));
+        BalanceSnapshot.Snapshot memory lonStakingLon = BalanceSnapshot.take(address(lonStaking), address(lon));
+        BalanceSnapshot.Snapshot memory userXLon = BalanceSnapshot.take(user, address(lonStaking));
+
+        uint256 veLonTotalPower = veLon.totalSupply();
```
Comment (charlesjhongc) [inline contracts/test/LONStaking.t.sol:1193]: unused?

File: contracts/test/LONStaking.t.sol [hunk 7]
```diff
@@ -1156,6 +1159,95 @@ contract LONStakingTest is Test {
...
+
+        // disable the conversion
+        lonStaking.disableConversion();
+
+        assertEq(address(lonStaking.veLon()), address(0x0));
+        assertEq(lonStaking.conversion(), false);
+    }
+
+    function _convertXLonToVeLonAndValidate(uint256 convertAmount, uint256 convertDuration) internal {
+        _stake(user, convertAmount);
```
Comment (charlesjhongc) [inline contracts/test/LONStaking.t.sol:1186]: I think `stake()` should be handled before calling `_convertXLonToVeLonAndValidate`.

**Review by charlesjhongc — COMMENTED** [review]
I think the `xxLon` should be renamed using a formal one. Perhaps `StakeMigrate` and `IStakeMigrate` ?

File: contracts/veLON.sol [hunk 8]
```diff
@@ -43,6 +46,39 @@ contract veLON is IveLON, ERC721, Ownable, ReentrancyGuard {
...
+    function disableConversion() external override onlyOwner {
+        conversion = false;
+        dstToken = address(0x0);
+        earlyWithdrawPenaltyRate = 3000;
+    }
+
+    /// @notice Help veLON holders convert their veLON to xxxLON.
+    function convertVeLontoXXXLon(bytes calldata _encodeData) external override {
+        require(conversion, "conversion is not enabled");
+        earlyWithdrawPenaltyRate = 0;
```
[the remark this answers is not in the document]
Comment (zeqing-guo) [reply contracts/veLON.sol:67]: We need to ensure `earlyWithdrawPenaltyRate` be zero before the conversion. Maybe `require(earlyWithdrawPenaltyRate == 0)` is a better choice.

File: contracts/interfaces/IveLON.sol [hunk 9]
```diff
@@ -40,6 +42,12 @@ interface IveLON is IERC721, IERC721Metadata {
 
     function createLock(uint256 _value, uint256 _lockDuration) external returns (uint256);
 
+    function createLockFor(
```
[the remark this answers is not in the document]
Comment (zeqing-guo) [reply contracts/interfaces/IveLON.sol:45]: It helps other contracts mint veLON for users. For example, when we upgrade xLON to veLON, lonStaking.sol will withdraw LON from staking contract and call `createFor` to create a veLON for `msg.sender`.

File: contracts/test/veLON/Convert.t.sol [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        _convertVeLontoXXXLonAndValidate(user, DEFAULT_STAKE_AMOUNT);
+    }
+
+    function _convertVeLontoXXXLonAndValidate(address staker, uint256 stakeAmount) public {
+        MockxxxLon xxxLon = new MockxxxLon(address(lon));
+        BalanceSnapshot.Snapshot memory veLonLon = BalanceSnapshot.take(address(veLon), address(lon));
+        BalanceSnapshot.Snapshot memory xxxLonLon = BalanceSnapshot.take(address(xxxLon), address(lon));
```
[the remark this answers is not in the document]
Comment (zeqing-guo) [reply contracts/test/veLON/Convert.t.sol:43]: The balance of `veLON` is meaningless after conversion, and `mockLON` does not have a method of `balanceOf` as the future implementation of `mochLON`. Thus, I have not check them in the function.

File: contracts/test/LONStaking.t.sol [hunk 11]
```diff
@@ -1156,6 +1159,95 @@ contract LONStakingTest is Test {
...
+        vm.assume(convertAmount <= lon.cap());
+        vm.assume(convertAmount.add(lon.totalSupply()) <= lon.cap());
+        vm.assume(convertDuration >= 1 weeks);
+        vm.assume(convertDuration <= 365 days);
+
+        lon.mint(user, convertAmount);
+        _convertXLonToVeLonAndValidate(convertAmount, convertDuration);
+    }
+
+    function testCannotConvert() public {
```
[the remark this answers is not in the document]
Comment (zeqing-guo) [reply contracts/test/LONStaking.t.sol:1241]: Haha, good idea!

File: contracts/veLON.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (charlesjhongc) [reply contracts/veLON.sol:67]: Yeah I think `require` would be more appropriate.

File: contracts/interfaces/IveLON.sol (hunk 9 above)
[the remark this answers is not in the document]
Comment (charlesjhongc) [reply contracts/interfaces/IveLON.sol:45]: Oh right, I got confused somehow.

File: contracts/test/veLON/Convert.t.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (charlesjhongc) [reply contracts/test/veLON/Convert.t.sol:43]: I mean, it should verify whether the token is actually `transferFrom`ed or not.

[the remark this answers is not in the document]
Comment (charlesjhongc) [reply contracts/test/veLON/Convert.t.sol:43]: I misunderstood.

File: contracts/test/veLON/Convert.t.sol [hunk 12]
```diff
@@ -3,23 +3,29 @@ pragma solidity 0.7.6;
...
+import "contracts/test/mocks/MockMigrateStake.sol";
 
 contract TestVeLONDeposit is TestVeLON {
     using BalanceSnapshot for BalanceSnapshot.Snapshot;
     using SafeMath for uint256;
 
     function testEnableAndDisableConversion() public {
-        MockxxxLon xxxLon = new MockxxxLon(address(lon));
+        MockMigrateStake migrateStake = new MockMigrateStake(address(lon));
+        _stakeAndValidate(user, DEFAULT_STAKE_AMOUNT, DEFAULT_LOCK_TIME);
```
Comment (charlesjhongc) [inline contracts/test/veLON/Convert.t.sol:14]: Minor: since `enabled` is checked at the very beginning of `convert()`, no need to stake in this test.

**Review by charlesjhongc — APPROVED** [review]

File: contracts/test/veLON/Convert.t.sol (hunk 12 above)
Comment (zeqing-guo) [reply contracts/test/veLON/Convert.t.sol]: Repositioned in [0ffca15](https://github.com/consenlabs/tokenlon-contracts/pull/70/commits/0ffca157bf0cf7cea5483e7f9b2e45b332e3aab2)

**Review by charlesjhongc — APPROVED** [review]

**Pull request merged**